### PR TITLE
feat(calling): move apis, listeners from calling client to line

### PR DESCRIPTION
**Pull request by Shreyas281299**
# COMPLETES #[SPARK-428389](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-428389)

## This pull request addresses
Required changes for line class.
1. Move below from callingClient to line: 
     - makeCall()
     - incomingCallListener()
     - getCallManager()

2. Add below new APIs in calling client and line:

CallingClient:
    - getActiveCalls()
    - getConnectedCall()
Line: 
    - getCall()

3. Change call event from callingClient:incomingCall to line:incomingCall.
4. To store lineData in the callManager. Added lineDict in call manager that holds the {deviceId: lineId} information. Also added addLine and getLine api to populate and fetch the lineDict
5. Changes in the samples page 
    - event handler for line:incoming_call event
    - changes from callingClient.makeCall to line.makeCall

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (arun3528) [inline packages/calling/src/CallingClient/CallingClient.ts:449]: i am little confused with the logic here , 

Why does someone want to receive all active calls in all line ?

if that's the intention do you think we need to maintain the keys of those calls in a array when ever the call gets active , any thoughts

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (arun3528) [inline packages/calling/src/CallingClient/CallingClient.ts:442]: needs some description for the function

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (arun3528) [inline packages/calling/src/CallingClient/CallingClient.ts:461]: same here need comments 

do you think its better to take line id as a parameter

**Review by arun3528 — CHANGES_REQUESTED** [review]

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.ts:445]: @arun3528 The purpose of creating this API was that if the user wants to fetch the information regarding which call is in a connected state at any point in time, they can call this API and get the call object which will have lineId as well.
Now they have this information on which line which call is in the connected state. Hence we need not provide lineId as a parameter here.

If we take lineId as a parameter here then the user has to keep providing each lineId one by one, calling this API multiple times and find which call is in the connected state.

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.ts:433]: @arun3528 So the structure this API will return is in the form of key-value where the key is the lineId and the value is an array of the calls active on that line.
The intention here is to just return that data back to the user if they want to fetch the information of all the calls that are active. When we say active here, it just means call is present but it can be in any state.

File: packages/calling/src/CallingClient/line/index.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/line/index.ts:236]: We don't need this comment here, please remove it

File: packages/calling/src/CallingClient/calling/types.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/calling/types.ts:244]: updateLine would be a better name for this internal function

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/calling/callManager.ts:269]: we can directly fetch lineId here instead of line object

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/calling/callManager.ts:437]: this should be a private function

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/calling/callManager.ts:429]: This also should be private

File: packages/calling/src/CallingClient/line/line.test.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/line/line.test.ts:181]: For successful cases, we should have a check for createCallSpy getting invoked with rights et of arguments. Make use of toBeCalledOnceWith

File: packages/calling/src/CallingClient/line/line.test.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/line/line.test.ts:231]: replace toBeCalledTimes(0) with not.toBeCalledOncewith

**Review by Kesari3008 — CHANGES_REQUESTED** [review]
I can see unused import in few files, please check and remove them

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts:420]: Added.

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts:439]: Added description

File: packages/calling/src/Events/types.ts
Comment (Shreyas281299) [reply packages/calling/src/Events/types.ts]: Changed.

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts:443]: For a held call, isConnected returns true and isHeld returns false. That's why both the conditions are required.

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts:441]: Changed

Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts:441]: this.callManager.getActiveCalls returns all the calls on every line.

File: packages/calling/src/CallingClient/CallingClient.test.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.test.ts]: Removed.

File: packages/calling/src/CallingClient/CallingClient.test.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.test.ts:622]: Added another call for testing

File: packages/calling/src/Events/types.ts
Comment (Shreyas281299) [reply packages/calling/src/Events/types.ts]: Changed all the event names

File: packages/calling/src/CallingClient/line/index.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/line/index.ts]: Removed

File: packages/calling/src/CallingClient/calling/types.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/calling/types.ts]: Changed the name

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/calling/callManager.ts]: Changed the function getLine to return only lineId

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/calling/callManager.ts]: In register function in line class, we are calling this function.

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/calling/callManager.ts:429]: In getConnectedCall and getActiveCalls in CallingClient we are calling this function

File: packages/calling/src/CallingClient/line/line.test.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/line/line.test.ts:181]: Added check for createCall

File: packages/calling/src/CallingClient/line/line.test.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/line/line.test.ts]: Changed.

File: packages/calling/src/CallingClient/calling/call.ts
Comment (BhargavSatya) [inline packages/calling/src/CallingClient/calling/call.ts:195]: can you keep lineId after deviceId.

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (BhargavSatya) [inline packages/calling/src/CallingClient/CallingClient.ts:423]: Doesn't lineDict has lineId as a key
```suggestion
    Object.keys(this.lineDict).forEach((lineId) => {
```

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (BhargavSatya) [inline packages/calling/src/CallingClient/CallingClient.ts:427]: Don't we have a getter on `lineId`? Is `lineId` a public attribute?

```suggestion
        if (calls[correlationId].getLineId() === lineId) {
```

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (BhargavSatya) [inline packages/calling/src/CallingClient/CallingClient.ts:426]: This =looks complicated to read.
```suggestion
      for (correlationId in calls) {
```

File: packages/calling/src/CallingClient/calling/call.ts
Comment (BhargavSatya) [inline packages/calling/src/CallingClient/calling/call.ts:2666]: Please add lineId in the description

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (BhargavSatya) [reply packages/calling/src/CallingClient/CallingClient.ts:423]:
```suggestion
  Object.keys(calls).forEach((correlationId) => {
      const call = calls[correlationId];
      if (!Object.keys(activeCalls).includes(call.lineId)) {
        activeCalls[call.lineId]  = [];
      } 
      activeCalls[call.lineId].append(call);
      
    });
```

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (BhargavSatya) [reply packages/calling/src/CallingClient/CallingClient.ts:420]: @Kesari3008  Can we have an optional parameter `lineId` which will only return calls from a a single line?

```suggestion
  public getActiveCalls(lineId?: string): Record<string, ICall[]> {
```

Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.ts:420]: Yeah, we can do that I guess. Let's discuss on it Monday

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.ts:443]: I am not sure if we should keep it like that, ideally isConnected should be false if the call is put on hold

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.ts:441]: It's only changed in getActiveCalls(). Change it in getConnectedCall as well

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/CallingClient.ts:437]: Single call object is returned so please correct the same in decription

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/calling/callManager.ts:444]: Change the function name also as per what is being returned

File: packages/calling/src/CallingClient/line/line.test.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/line/line.test.ts:197]: In this case and the below cases also where the call object is truthy, won't createCall will be invoked? If yes, please add the spy invocation check in those as well

File: packages/calling/src/CallingClient/CallingClient.test.ts
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.test.ts:622]: I still see only one call here

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/calling/callManager.ts:434]: Avoid adding @param and @returns in the function definition as they are only for js file not ts. Just plain text describing the purpose of function should be enough.

**Review by Kesari3008 — CHANGES_REQUESTED** [review]

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts]: Changed in getConnectedCall

File: packages/calling/src/CallingClient/calling/call.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/calling/call.ts]: Moved lineId after deviceId

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts]: Changed to the suggested changes to make the code more readable

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts]: We dont have a getter for lineId. Yes, lineId is a public attribute.

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts]: Changed

File: packages/calling/src/CallingClient/calling/call.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/calling/call.ts]: Added

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.ts]: Changed

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/calling/callManager.ts]: Changed.

File: packages/calling/src/CallingClient/line/line.test.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/line/line.test.ts:197]: Yes createCall will be invoked. Added checks for createCall

File: packages/calling/src/CallingClient/calling/callManager.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/calling/callManager.ts]: Removed @params.

File: packages/calling/src/CallingClient/CallingClient.ts
Comment (Kesari3008) [reply packages/calling/src/CallingClient/CallingClient.ts:420]: We will do this as part of separate PR and discuss the approach for the same

File: packages/calling/src/CallingClient/CallingClient.test.ts
Comment (Kesari3008) [inline packages/calling/src/CallingClient/CallingClient.test.ts:635]: Shouldn't the state here be S_CALL_HOLD ?

**Review by Kesari3008 — APPROVED** [review]
Please take care of below points:

1. One minor comment in UT file.
2. isConnected state needs to be corrected so let's make a note of it. 
3. Ensure all the conflicts are resolved before merge and all UT's are passing.
4. Remove extra imports if there are any present.
5. Found one UT in callManager.test.ts file showing error for localAudioTrack being passed to call.answer(). Please fix that before merge.

**Review by BhargavSatya — APPROVED** [review]

**Comment by BhargavSatya** [comment]
Please take care of Priya's approval comments.

File: packages/calling/src/CallingClient/CallingClient.test.ts
Comment (Shreyas281299) [reply packages/calling/src/CallingClient/CallingClient.test.ts]: Yes, it should be S_CALL_HOLD. Changed

**Pull request merged**
